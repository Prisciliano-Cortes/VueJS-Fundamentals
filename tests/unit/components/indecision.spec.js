import { shallowMount } from "@vue/test-utils";
import Indecision from '@/components/Indecision.vue'

describe('Indecision component', () => {

    let wrapper;
    let clgSpy;

    global.fetch = jest.fn( () => Promise.resolve({
        json: () => Promise.resolve({
            answer: 'yes',
            forced: false,
            image:  'https://yesno.wtf/assets/yes/2.gif'
        })
    }))

    beforeEach( () => {
        wrapper = shallowMount( Indecision )

        clgSpy = jest.spyOn( console, 'log')

        jest.clearAllMocks();
    })

    test('You must match the snapshot', () => {

        expect( wrapper.html() ).toMatchSnapshot()
    })

    test('Writing to input should not trigger anything', async() => {

        const getAnswerSpy = jest.spyOn( wrapper.vm, 'getAnswer' )

        const input = wrapper.find("input")

        await input.setValue('Hello world')

        expect( clgSpy ).toHaveBeenCalled()
        expect( clgSpy ).toHaveBeenCalledTimes(1)

        expect( getAnswerSpy ).not.toHaveBeenCalled()
    })

    test('Write ? must shoot the getAnswer', async() => {
        const getAnswerSpy = jest.spyOn( wrapper.vm, 'getAnswer' )

        const input = wrapper.find('input')
        await input.setValue('Hello world?')

        expect( clgSpy ).toHaveBeenCalledTimes(1)
        expect( getAnswerSpy ).toHaveBeenCalled()
    })

    test('Test in getAnswer', async() => {
        await wrapper.vm.getAnswer()

        const img = wrapper.find('img')

        expect( img.exists() ).toBeTruthy()

        expect( wrapper.vm.img ).toBe('https://yesno.wtf/assets/yes/2.gif')

        expect( wrapper.vm.answer ).toBe('SI')
    })

    test('Test in getAnswer - failed API', async() => {
        fetch.mockImplementationOnce( () => Promise.reject('API is down') )

        await wrapper.vm.getAnswer()

        const img = wrapper.find('img')

        expect( img.exists() ).toBeFalsy()
        
        expect( wrapper.vm.answer ).toBe('Failed to load from API')
    })
})