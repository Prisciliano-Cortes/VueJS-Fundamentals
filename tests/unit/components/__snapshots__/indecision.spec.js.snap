// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indecision component You must match the snapshot 1`] = `
<!--v-if-->
<div class="bg-dark"></div>
<div class="indecision-container"><input type="text" placeholder="Write your a question">
  <p>Remember finish your question with ?</p>
  <!--v-if-->
</div>
`;
